File: backend/api/assets/starters/cpp-linkedlist/memo/LinkedList.cpp
#include "LinkedList.hpp"

LinkedList::LinkedList() : head_(nullptr), tail_(nullptr), size_(0) {}
LinkedList::LinkedList(const LinkedList &other) : LinkedList() { copyFrom(other); }

LinkedList::LinkedList(LinkedList &&other) noexcept
    : head_(other.head_), tail_(other.tail_), size_(other.size_)
{
    other.head_ = other.tail_ = nullptr;
    other.size_ = 0;
}

LinkedList &LinkedList::operator=(const LinkedList &other)
{
    if (this != &other)
    {
        clear();
        copyFrom(other);
    }
    return *this;
}
LinkedList &LinkedList::operator=(LinkedList &&other) noexcept
{
    if (this != &other)
    {
        clear();
        head_ = other.head_;
        tail_ = other.tail_;
        size_ = other.size_;
        other.head_ = other.tail_ = nullptr;
        other.size_ = 0;
    }
    return *this;
}
LinkedList::~LinkedList() { clear(); }

bool LinkedList::empty() const noexcept { return size_ == 0; }
std::size_t LinkedList::size() const noexcept { return size_; }

void LinkedList::clear() noexcept
{
    for (auto *n = head_; n;)
    {
        auto *next = n->next;
        delete n;
        n = next;
    }
    head_ = tail_ = nullptr;
    size_ = 0;
}

void LinkedList::push_front(int v)
{
    auto *n = new Node(v);
    n->next = head_;
    head_ = n;
    if (!tail_)
        tail_ = n;
    ++size_;
}
void LinkedList::push_back(int v)
{
    auto *n = new Node(v);
    if (tail_)
    {
        tail_->next = n;
        tail_ = n;
    }
    else
    {
        head_ = tail_ = n;
    }
    ++size_;
}
bool LinkedList::pop_front(int &out)
{
    if (!head_)
        return false;
    auto *n = head_;
    out = n->value;
    head_ = n->next;
    if (!head_)
        tail_ = nullptr;
    delete n;
    --size_;
    return true;
}

int &LinkedList::front() { return head_->value; }
const int &LinkedList::front() const { return head_->value; }
int &LinkedList::back() { return tail_->value; }
const int &LinkedList::back() const { return tail_->value; }

bool LinkedList::insert(std::size_t index, int value)
{
    if (index > size_)
        return false;
    if (index == 0)
    {
        push_front(value);
        return true;
    }
    if (index == size_)
    {
        push_back(value);
        return true;
    }
    auto *prev = head_;
    for (std::size_t i = 0; i + 1 < index; ++i)
        prev = prev->next;
    auto *n = new Node(value);
    n->next = prev->next;
    prev->next = n;
    ++size_;
    return true;
}
bool LinkedList::erase(std::size_t index)
{
    if (index >= size_)
        return false;
    if (index == 0)
    {
        int tmp;
        return pop_front(tmp);
    }
    auto *prev = head_;
    for (std::size_t i = 0; i + 1 < index; ++i)
        prev = prev->next;
    auto *victim = prev->next;
    prev->next = victim->next;
    if (victim == tail_)
        tail_ = prev;
    delete victim;
    --size_;
    return true;
}

std::vector<int> LinkedList::toVector() const
{
    std::vector<int> out;
    out.reserve(size_);
    for (auto *n = head_; n; n = n->next)
        out.push_back(n->value);
    return out;
}
void LinkedList::copyFrom(const LinkedList &other)
{
    for (auto *n = other.head_; n; n = n->next)
        push_back(n->value);
}

File: backend/api/assets/starters/java-linkedlist/memo/LinkedList.java
import java.util.*;

public class LinkedList {
	private static class Node {
		int value;
		Node next;

		Node(int v) {
			value = v;
		}
	}

	public static class IntBox {
		public int value;
	}

	private Node head, tail;
	private int size;

	public LinkedList() {
	}

	public LinkedList(LinkedList other) {
		copyFrom(other);
	}

	// Simulate move-ctor by transferring links
	public static LinkedList moveFrom(LinkedList src) {
		LinkedList dst = new LinkedList();
		dst.head = src.head;
		dst.tail = src.tail;
		dst.size = src.size;
		src.head = src.tail = null;
		src.size = 0;
		return dst;
	}

	// Simulate move-assign
	public void moveAssignFrom(LinkedList src) {
		if (this == src)
			return;
		clear();
		this.head = src.head;
		this.tail = src.tail;
		this.size = src.size;
		src.head = src.tail = null;
		src.size = 0;
	}

	public boolean isEmpty() {
		return size == 0;
	}

	public int size() {
		return size;
	}

	public void clear() {
		Node n = head;
		while (n != null) {
			Node next = n.next;
			n.next = null;
			n = next;
		}
		head = tail = null;
		size = 0;
	}

	public void pushFront(int v) {
		Node n = new Node(v);
		n.next = head;
		head = n;
		if (tail == null)
			tail = n;
		size++;
	}

	public void pushBack(int v) {
		Node n = new Node(v);
		if (tail != null) {
			tail.next = n;
			tail = n;
		} else {
			head = tail = n;
		}
		size++;
	}

	public boolean popFront(IntBox out) {
		if (head == null)
			return false;
		Node n = head;
		if (out != null)
			out.value = n.value;
		head = n.next;
		if (head == null)
			tail = null;
		size--;
		return true;
	}

	public int front() {
		if (head == null)
			throw new IllegalStateException("empty");
		return head.value;
	}

	public int back() {
		if (tail == null)
			throw new IllegalStateException("empty");
		return tail.value;
	}

	public boolean insert(int index, int value) {
		if (index < 0 || index > size)
			return false;
		if (index == 0) {
			pushFront(value);
			return true;
		}
		if (index == size) {
			pushBack(value);
			return true;
		}
		Node prev = head;
		for (int i = 0; i + 1 < index; i++)
			prev = prev.next;
		Node n = new Node(value);
		n.next = prev.next;
		prev.next = n;
		size++;
		return true;
	}

	public boolean erase(int index) {
		if (index < 0 || index >= size)
			return false;
		if (index == 0) {
			IntBox box = new IntBox();
			return popFront(box);
		}
		Node prev = head;
		for (int i = 0; i + 1 < index; i++)
			prev = prev.next;
		Node victim = prev.next;
		prev.next = victim.next;
		if (victim == tail)
			tail = prev;
		size--;
		return true;
	}

	public List<Integer> toList() {
		ArrayList<Integer> out = new ArrayList<>(size);
		for (Node n = head; n != null; n = n.next)
			out.add(n.value);
		return out;
	}

	private void copyFrom(LinkedList other) {
		for (int v : other.toList())
			pushBack(v);
	}
}
